<a href="https://colab.research.google.com/github/Jolanta27/pandas_exercises/blob/master/wind_data_analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from google.colab import files
uploaded = files.upload()

Saving wind.csv to wind.csv


1. Wczytaj dane z załączonego pliku .csv i zapisz je w zmiennej data klasy pd.DataFrame(). Zastąp pierwsze 3 kolumny odpowiednim formatem datetime.

Wszystkie niezbędne biblioteki są importowane.

Pierwsza kolumna zawiera właściwe dane po odczytaniu pliku .csv w pandas.

Zaimportowana ramka danych zawiera 6574 wierszy i 13 kolumn.


In [22]:
import pandas as pd
data = pd.read_csv('wind.csv')

data['Yr_Mo_Dy'] = pd.to_datetime(data[['Yr', 'Mo', 'Dy']].astype(str).agg('-'.join, axis=1), format='%y-%m-%d')

future = data['Yr_Mo_Dy'] > pd.to_datetime('today')
data.loc[future, 'Yr_Mo_Dy'] -= pd.DateOffset(years=100)

data = data.drop(['Yr', 'Mo', 'Dy'], axis=1)

print(data.head)
print(data.shape)

<bound method NDFrame.head of         RPT    VAL    ROS    KIL    SHA    BIR    DUB    CLA    MUL    CLO  \
0     15.04  14.96  13.17   9.29    NaN   9.87  13.67  10.25  10.83  12.58   
1     14.71    NaN  10.83   6.50  12.62   7.67  11.50  10.04   9.79   9.67   
2     18.50  16.88  12.33  10.13  11.17   6.17  11.25    NaN   8.50   7.67   
3     10.58   6.63  11.75   4.58   4.54   2.88   8.63   1.79   5.83   5.88   
4     13.33  13.25  11.42   6.17  10.71   8.21  11.92   6.54  10.92  10.34   
...     ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
6569  17.58  16.96  17.62   8.08  13.21  11.67  14.46  15.59  14.04  14.00   
6570  13.21   5.46  13.46   5.00   8.12   9.42  14.33  16.25  15.25  18.05   
6571  14.00  10.29  14.42   8.71   9.71  10.54  19.17  12.46  14.50  16.42   
6572  18.50  14.04  21.29   9.13  12.75   9.71  18.08  12.87  12.46  12.12   
6573  20.33  17.41  27.29   9.59  12.08  10.13  19.25  11.63  11.58  11.38   

        BEL    MAL   Yr_Mo_Dy  
0

2. Rok 2061? Z pewnością nie posiadamy danych z przyszłości. Skonstruuj funkcję korygującą tę anomalię i zaimplementuj ją. Jeśli rok daty przekracza 1989, oznacza to błąd danych. Sprawdź, jak często się to zdarza. W takich przypadkach odejmij 100 od roku, aby naprawić błąd.

Wszystkie daty przypadają na lata 1961-1978

In [43]:
import pandas as pd

def correct_year(date):
    if date.year > 1989:
        return date.replace(year = date.year - 100)
    else:
        return date

data = pd.read_csv('wind.csv', parse_dates=[['Yr', 'Mo', 'Dy']])

data['Yr_Mo_Dy'] = data['Yr_Mo_Dy'].apply(correct_year)

anomaly_count = sum(data['Yr_Mo_Dy'].dt.year > 1989)

print(f"The anomaly occurs {anomaly_count} times.")

print(data.head())


<ipython-input-43-ecc7bf0e4788>:9: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  data = pd.read_csv('wind.csv', parse_dates=[['Yr', 'Mo', 'Dy']])


The anomaly occurs 0 times.
    Yr_Mo_Dy    RPT    VAL    ROS    KIL    SHA   BIR    DUB    CLA    MUL  \
0 1961-01-01  15.04  14.96  13.17   9.29    NaN  9.87  13.67  10.25  10.83   
1 1961-01-02  14.71    NaN  10.83   6.50  12.62  7.67  11.50  10.04   9.79   
2 1961-01-03  18.50  16.88  12.33  10.13  11.17  6.17  11.25    NaN   8.50   
3 1961-01-04  10.58   6.63  11.75   4.58   4.54  2.88   8.63   1.79   5.83   
4 1961-01-05  13.33  13.25  11.42   6.17  10.71  8.21  11.92   6.54  10.92   

     CLO    BEL    MAL  
0  12.58  18.50  15.04  
1   9.67  17.54  13.83  
2   7.67  12.75  12.71  
3   5.88   5.46  10.88  
4  10.34  12.92  11.83  


3. Ustaw poprawione daty jako indeksy. Zwróć uwagę na typ danych, powinien to być datetime64!

Typ kolumny daty został przekształcony do datetime64.

Kolumna Date jest ustawiona jako indeks DataFrame

In [44]:
data = pd.read_csv('wind.csv', parse_dates=[['Yr', 'Mo', 'Dy']])

# Zastosuj funkcję correct_year do kolumny 'Yr_Mo_Dy'
data['Yr_Mo_Dy'] = data['Yr_Mo_Dy'].apply(correct_year)

# Ustaw poprawione daty jako indeks
data.set_index('Yr_Mo_Dy', inplace=True)

# Sprawdź typ danych indeksu
print(data.index.dtype)

print(data.head())

<ipython-input-44-81e4d7e6c650>:1: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  data = pd.read_csv('wind.csv', parse_dates=[['Yr', 'Mo', 'Dy']])


datetime64[ns]
              RPT    VAL    ROS    KIL    SHA   BIR    DUB    CLA    MUL  \
Yr_Mo_Dy                                                                   
1961-01-01  15.04  14.96  13.17   9.29    NaN  9.87  13.67  10.25  10.83   
1961-01-02  14.71    NaN  10.83   6.50  12.62  7.67  11.50  10.04   9.79   
1961-01-03  18.50  16.88  12.33  10.13  11.17  6.17  11.25    NaN   8.50   
1961-01-04  10.58   6.63  11.75   4.58   4.54  2.88   8.63   1.79   5.83   
1961-01-05  13.33  13.25  11.42   6.17  10.71  8.21  11.92   6.54  10.92   

              CLO    BEL    MAL  
Yr_Mo_Dy                         
1961-01-01  12.58  18.50  15.04  
1961-01-02   9.67  17.54  13.83  
1961-01-03   7.67  12.75  12.71  
1961-01-04   5.88   5.46  10.88  
1961-01-05  10.34  12.92  11.83  


4. Oblicz, ile wartości brakuje dla każdej lokalizacji w całym okresie.

Liczba brakujących wartości dla każdej lokalizacji:


In [45]:
missing_values = data.isnull().sum()

print(missing_values)

RPT    6
VAL    3
ROS    2
KIL    5
SHA    2
BIR    0
DUB    3
CLA    2
MUL    3
CLO    1
BEL    0
MAL    4
dtype: int64


5. Oblicz średnią prędkość wiatru we wszystkich lokalizacjach i dla całego zbioru danych (innymi słowy, z każdego wiersza jednocześnie). Zaokrąglij ostateczną wartość do dwóch miejsc po przecinku.

Oczekiwany wynik to około: 10.23

In [48]:
average_wind_speed = data.mean().mean()
average_wind_speed = round(average_wind_speed, 2)

print(average_wind_speed)

10.23


6. Skonstruuj pd.DataFrame() o nazwie wind_stats i określ minimalne, maksymalne, średnie prędkości wiatru oraz odchylenia standardowe prędkości wiatru dla każdej lokalizacji we wszystkich dniach.

In [50]:
wind_stats = data.describe()
wind_stats = wind_stats.loc[['min', 'max', 'mean', 'std']]

print(wind_stats)

            RPT        VAL        ROS        KIL        SHA        BIR  \
min    0.670000   0.210000   1.500000   0.000000   0.130000   0.000000   
max   35.800000  33.370000  33.840000  28.460000  37.540000  26.160000   
mean  12.362987  10.644314  11.660526   6.306468  10.455834   7.092254   
std    5.618413   5.267356   5.008450   3.605811   4.936125   3.968683   

            DUB        CLA        MUL        CLO        BEL        MAL  
min    0.000000   0.000000   0.000000   0.040000   0.130000   0.670000  
max   30.370000  31.080000  25.880000  28.210000  42.380000  42.540000  
mean   9.797343   8.495053   8.493590   8.707332  13.121007  15.599079  
std    4.977555   4.499449   4.166872   4.503954   5.835037   6.699794  


7. Skonstruuj pd.DataFrame() o nazwie wind_stats_daily i oblicz minimalną, maksymalną, średnią prędkość wiatru i odchylenia standardowe prędkości wiatru we wszystkich lokalizacjach dla każdego dnia osobno.

Wynikowa tabela powinna wyglądać następująco:

| min | max | średnia | std | |------------|------|-------|-----------|----------| | Yr_Mo_Dy | | | | | | | | 1961-01-01 | 9.29 | 18.50 | 13.018182 | 2.808875 | | 1961-01-02 | 6.50 | 17.54 | 11.336364 | 3.188994 | | 1961-01-03 | 6.17 | 18.50 | 11.641818 | 3.681912 | | 1961-01-04 | 1.79 | 11.75 | 6.619167 | 3.198126 | | 1961-01-05 | 6.17 | 13.33 | 10.630000 | 2.445356 |

In [51]:
wind_stats_daily = data.resample('D').agg(['min', 'max', 'mean', 'std'])

print(wind_stats_daily.head())

              RPT                      VAL                      ROS         \
              min    max   mean std    min    max   mean std    min    max   
Yr_Mo_Dy                                                                     
1961-01-01  15.04  15.04  15.04 NaN  14.96  14.96  14.96 NaN  13.17  13.17   
1961-01-02  14.71  14.71  14.71 NaN    NaN    NaN    NaN NaN  10.83  10.83   
1961-01-03  18.50  18.50  18.50 NaN  16.88  16.88  16.88 NaN  12.33  12.33   
1961-01-04  10.58  10.58  10.58 NaN   6.63   6.63   6.63 NaN  11.75  11.75   
1961-01-05  13.33  13.33  13.33 NaN  13.25  13.25  13.25 NaN  11.42  11.42   

            ...    CLO        BEL                      MAL                    
            ...   mean std    min    max   mean std    min    max   mean std  
Yr_Mo_Dy    ...                                                               
1961-01-01  ...  12.58 NaN  18.50  18.50  18.50 NaN  15.04  15.04  15.04 NaN  
1961-01-02  ...   9.67 NaN  17.54  17.54  17.54 NaN  13.83 

8. Określ średnią prędkość wiatru dla każdej lokalizacji w styczniu.

In [52]:
january_winds = data[data.index.month == 1].mean()

print(january_winds)

RPT    14.847325
VAL    12.914560
ROS    13.299624
KIL     7.199498
SHA    11.667734
BIR     8.054839
DUB    11.819355
CLA     9.512047
MUL     9.543208
CLO    10.053566
BEL    14.550520
MAL    18.028763
dtype: float64


9. Oblicz średnią roczną prędkość wiatru dla każdej lokalizacji.

In [53]:
annual_wind_speed = data.resample('A').mean()

print(annual_wind_speed)

                  RPT        VAL        ROS       KIL        SHA       BIR  \
Yr_Mo_Dy                                                                     
1961-12-31  12.299583  10.351796  11.362369  6.958227  10.881763  7.729726   
1962-12-31  12.246923  10.110438  11.732712  6.960440  10.657918  7.393068   
1963-12-31  12.813452  10.836986  12.541151  7.330055  11.724110  8.434712   
1964-12-31  12.363661  10.920164  12.104372  6.787787  11.454481  7.570874   
1965-12-31  12.451370  11.075534  11.848767  6.858466  11.024795  7.478110   
1966-12-31  13.461973  11.557205  12.020630  7.345726  11.805041  7.793671   
1967-12-31  12.737151  10.990986  11.739397  7.143425  11.630740  7.368164   
1968-12-31  11.835628  10.468197  11.409754  6.477678  10.760765  6.067322   
1969-12-31  11.166356   9.723699  10.902000  5.767973   9.873918  6.189973   
1970-12-31  12.600329  10.726932  11.730247  6.217178  10.567370  7.609452   
1971-12-31  11.273123   9.095178  11.088329  5.241507   9.440329

10. Zmniejsz zestaw danych do częstotliwości tygodniowej dla każdej lokalizacji, agregując punkty danych, aby reprezentowały każdy tydzień (uzyskując średnie wartości).

In [54]:
weekly_data = data.resample('W').mean()

print(weekly_data.head())

                  RPT        VAL        ROS        KIL        SHA        BIR  \
Yr_Mo_Dy                                                                       
1961-01-01  15.040000  14.960000  13.170000   9.290000        NaN   9.870000   
1961-01-08  13.541429  11.486667  10.487143   6.417143   9.474286   6.435714   
1961-01-15  12.468571   8.967143  11.958571   4.630000   7.351429   5.072857   
1961-01-22  13.204286   9.862857  12.982857   6.328571   8.966667   7.417143   
1961-01-29  19.880000  16.141429  18.225714  12.720000  17.432857  14.828571   

                  DUB        CLA        MUL        CLO        BEL        MAL  
Yr_Mo_Dy                                                                      
1961-01-01  13.670000  10.250000  10.830000  12.580000  18.500000  15.040000  
1961-01-08  11.061429   6.616667   8.434286   8.497143  12.481429  13.238571  
1961-01-15   7.535714   6.820000   5.712857   7.571429  11.125714  11.024286  
1961-01-22   9.257143   7.875714   7.145714 

11. Oblicz minimalne, maksymalne, średnie prędkości wiatru i odchylenia standardowe prędkości wiatru we wszystkich lokalizacjach dla każdego tygodnia, począwszy od tygodnia 2 stycznia 1961 r., obejmującego początkowe 52 tygodnie.

Wynik jest następujący:

| | RPT | | | VAL | | | | |------------|-------|-------|-----------|----------|------|-------|-----------|----------| | | min | max | śr | std | | min | max | śr | std | | | | | | | | | | | | | | | | | | | | | | | | | | | 1961-01-08 | 10.58 | 18.50 | 13.541429 | 2.631321 | 6.63 | 16.88 | 11.486667 | 3.949525 | | 1961-01-15 | 9.04 | 19.75 | 12.468571 | 3.555392 | 3.54 | 12.08 | 8.967143 | 3.148945 | | 1961-01-22 | 4.92 | 19.83 | 13.204286 | 5.337402 | 3.42 | 14.37 | 9.862857 | 3.837785 | | 1961-01-29 | 13.62 | 25.04 | 19.880000 | 4.619061 | 9.96 | 23.91 | 16.141429 | 5.170224 | | 1961-02-05 | 10.58 | 24.21 | 16.827143 | 5.251408 | 9.46 | 24.21 | 15.460000 | 5.187395 |

In [55]:
weekly_stats = data['1961-01-02':'1962-01-01'].resample('W').agg(['min', 'max', 'mean', 'std'])

print(weekly_stats.head())

              RPT                               VAL                    \
              min    max       mean       std   min    max       mean   
Yr_Mo_Dy                                                                
1961-01-08  10.58  18.50  13.541429  2.631321  6.63  16.88  11.486667   
1961-01-15   9.04  19.75  12.468571  3.555392  3.54  12.08   8.967143   
1961-01-22   4.92  19.83  13.204286  5.337402  3.42  14.37   9.862857   
1961-01-29  13.62  25.04  19.880000  4.619061  9.96  23.91  16.141429   
1961-02-05  10.58  24.21  16.827143  5.251408  9.46  24.21  15.460000   

                        ROS         ...        CLO              BEL         \
                 std    min    max  ...       mean       std    min    max   
Yr_Mo_Dy                            ...                                      
1961-01-08  3.949525   7.62  12.33  ...   8.497143  1.704941   5.46  17.54   
1961-01-15  3.148945   7.08  19.50  ...   7.571429  4.084293   5.25  20.71   
1961-01-22  3.837785   7.